sourcegraph/sourcegraph
# Cody: Inline Assist Feature - File Flow

![image](https://github.com/sourcegraph/sourcegraph/assets/68532117/3f76531a-40d4-4568-93c1-3c07aa78d8a3)

Add a feature to Cody Inline Assist that automatically creates a new file with tailored content based on the selected code. 

This was developed with the goal of preserving developer flow, which is one of the main goals of Non-Stop Cody. 

This feature addresses recent frustrations that I've encountered personally while writing new tests, and trying to improve the current code generation process. Unlike the previous approach that required manual specification of framework and details, this feature automates file creation and ensures higher quality code output. By eliminating extra steps and streamlining the process, developers can now enjoy a seamless and efficient code generation experience.

Feature highlights:
- Hidden behind the `inline` experimental chat
- Cody now receives additional context in the form of files from the user's current directory to provide accurate and relevant output. This ensures Cidy understands the user's framework, model, and style preferences before generating new content for the selected code.
- File creation:
  - If the user's query includes the word "test," the feature automatically creates a test file in the same directory. e.g `helpers.test.ts`
  - If a file with the same name exists, it will automatically insert to the end of the existing file
  - For queries without the word "test," the assistant adds "cody" between the file name and the extension, allowing users to identify the newly generated files easily. eg. `index.cody.ts`
- Once Cody has completed the request and added the content to the new file, it is instantly presented to the user. 

To use this feature,  add `/t` or `/touch` to the start of your input in the Inline-Assist chat / comment input box.

This seamless integration allows developers to continue their work uninterrupted, preserving their workflow and maximizing productivity.

Still deciding on naming it File Flow or File Touch 🤔 

## Test plan



How to use:

- Start Cody from this branch in VS Code
- make sure you have the `inline` experimental flag enabled
- Highlight code you want Cody to generate the file from
- Start a new Inline Assist comment thread
- Enter `/t` and then follow by your instruction for Cody to create new content with
- Wait for Cody's response. The new file will first be created with empty content. You can continue working in your editor as Cody will let you know once the file is ready.

Loom: https://www.loom.com/share/c132ffce708c497f97a5ed3167c21b2b

TODO: Will add a e2e test in a separate PR (DONE, included in this PR)

```sh
❯ pnpm test:e2e inline-assist-file-flow

> cody-ai@0.1.5 test:e2e /Users/bwork/dev/sourcegraph/client/cody
> npx playwright install && pnpm run --silent build:dev && playwright test "inline-assist-file-flow"

[0]
[0]   dist/extension.js      5.0mb ⚠️
[0]   dist/extension.js.map  7.6mb
[0]
[0] ⚡ Done in 183ms
[0] pnpm esbuild --sourcemap exited with code 0
[1] vite build --mode development exited with code 0

Running 1 test using 1 worker

  ✓  1 inline-assist-file-flow.test.ts:6:5 › start a fixup job from inline assist with valid auth (4.4s)
Found existing install in /Users/bwork/dev/sourcegraph/client/cody/.vscode-test/vscode-darwin-arm64-1.78.2. Skipping download
Mock server listening on port 49300
(node:86520) [DEP0147] DeprecationWarning: In future versions of Node.js, fs.rmdir(path, { recursive: true }) will be removed. Use fs.rm(path, { recursive: true }) instead
(Use `node --trace-deprecation ...` to show where the warning was created)

  1 passed (5.0s)
```
